profile-widget/profilewidget2.cpp

```diff
@@ -2044,70 +2045,75 @@ void ProfileWidget2::clearPictures()
 	pictures.clear();
 }
 
-void ProfileWidget2::updatePictures(const QModelIndex &from, const QModelIndex &to)
+void ProfileWidget2::updateThumbnail(QString filename, QImage thumbnail)
 {
-	DivePictureModel *m = DivePictureModel::instance();
-	for (int picNr = from.row(); picNr <= to.row(); ++picNr) {
-		int picItemNr = picNr - m->rowDDStart;
-		if (picItemNr < 0 || (size_t)picItemNr >= pictures.size())
-			return;
-		if (!pictures[picItemNr])
-			return;
-
-		pictures[picItemNr]->setPixmap(m->index(picNr, 0).data(Qt::UserRole).value<QPixmap>());
+	auto it = std::find_if(pictures.begin(), pictures.end(), [&filename](const PictureEntry &e)
+			       { return e.filename == filename; });
+	if (it != pictures.end() && it->thumbnail) {
+		int size = Thumbnailer::defaultThumbnailSize();
+		it->thumbnail->setPixmap(QPixmap::fromImage(thumbnail.scaled(size, size, Qt::KeepAspectRatio)));
 	}
 }
 
+bool ProfileWidget2::PictureEntry::operator< (const PictureEntry &e) const
+{
+	return std::tie(offset.seconds, filename) < std::tie(e.offset.seconds, e.filename);
+}
+
 void ProfileWidget2::plotPictures()
 {
-	DivePictureModel *m = DivePictureModel::instance();
-	pictures.resize(m->rowDDEnd - m->rowDDStart);
+	pictures.clear();
+	if (currentState == ADD || currentState == PLAN)
+		return;
+	FOR_EACH_PICTURE(current_dive)
+		pictures.push_back(PictureEntry{ picture->offset, QString(picture->filename),  {} });
+	if (pictures.empty())
+		return;
+	std::sort(pictures.begin(), pictures.end());
 
 	double x, y, lastX = -1.0, lastY = -1.0;
-	for (int i = m->rowDDStart; i < m->rowDDEnd; i++) {
-		int picItemNr = i - m->rowDDStart;
-		int offsetSeconds = m->index(i, 1).data(Qt::UserRole).value<int>();
-		// it's a correct picture, but doesn't have a timestamp: only show on the widget near the
-		// information area. A null pointer in the pictures array indicates that this picture is not
-		// shown.
-		if (!offsetSeconds) {
-			pictures[picItemNr].reset();
+	int size = Thumbnailer::defaultThumbnailSize();
+	for (PictureEntry &e: pictures) {
+		// Only show pictures whose offsets are withing the profile
+		if (e.offset.seconds <= 0 || e.offset.seconds >= current_dive->duration.seconds)
 			continue;
-		}
-		DivePictureItem *item = pictures[picItemNr].get();
-		if (!item) {
-			item = new DivePictureItem;
-			pictures[picItemNr].reset(item);
-			scene()->addItem(item);
-		}
-		item->setVisible(prefs.show_pictures_in_profile);
-		item->setPixmap(m->index(i, 0).data(Qt::UserRole).value<QPixmap>());
-		item->setFileUrl(m->index(i, 1).data().toString());
+		e.thumbnail.reset(new DivePictureItem);
```

Comment:
what does this do?

Reply:
This sets the `std::unique_ptr` to a newly allocated object.

I'm using `std::unique_ptr<DivePictureItem>` instead of `DivePictureItem *` here, because it makes it unnecessary to bother with memory management. Grow, shrink, clear the vector - it will take care that no memory is leaked.

I used `std::unique_ptr` instead of Qt's (subpar) version `QScopedPointer`, because the latter didn't work properly in a vector (no move semantics). But with the simplifications in this commit, perhaps `QScopedPointer` works again. I can give it a try.